okio/src/nonJvmMain/kotlin/okio/internal/AbstractMessageDigest.kt

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (C) 2018 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package okio.internal
+
+import okio.OkioMessageDigest
+
+internal abstract class AbstractMessageDigest : OkioMessageDigest {
+
+  private var messageLength: Long = 0
+  private var unprocessed = Bytes.EMPTY
+  protected abstract var currentDigest: HashDigest
+
+  override fun update(input: ByteArray) {
+    for (chunk in (unprocessed + input.toBytes()).chunked(64)) {
+      when (chunk.size) {
+        64 -> {
+          currentDigest = processChunk(chunk, currentDigest)
+          messageLength += 64
+        }
+        else -> unprocessed = chunk
+      }
+    }
+  }
+
+  override fun digest(): ByteArray {
+    val finalMessageLength = messageLength + unprocessed.size
+
+    val finalMessage = byteArrayOf(
+      *unprocessed.toByteArray(),
+      0x80.toByte(),
+      *ByteArray(((56 - (finalMessageLength + 1) % 64) % 64).toInt()),
```

Comment:
what’s the size of this array if `finalMessageLength` is 56 ? Ie. if we attempt to process a single 56-element array?

Reply:
The resulting array would be 63 bytes. Given `finalMessageLength` is 55 bytes, the resulting array would be 0 bytes. 

Written more verbosely, the expression in question can be written as the following.

```kotlin
while (finalMessageLength != 56) {
    finalMessage.append(0.toByte())
}
```

We do this because the message needs to be padded with 0 bytes such that we have 8 bytes left over for the length of the message to be appended in byte representation. 

The wikipedia explanation might make more sense than my explanation.

> append 0 ≤ k < 512 bits '0', such that the resulting message length in bits is congruent to −64 ≡ 448 (mod 512)

Reply:
I think it's gonna get -1 instead of 63 here.

Reply:
Aaaaaaaand you're absolutely correct. Will fix in update.

Reply:
You can probably find a way to do `% 64` only once!

Reply:
Yeah, did some research. The python implementation of mod (which i was using as a reference) by default has absolute modulus and instead returns the positive complement. 

I have an `absoluteMod` function written now which seems to do the job. Will write an accompanying test too :)

Reply:
Fixed :)